driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java

```diff
@@ -405,31 +414,47 @@ private boolean handleMongoWriteConcernWithResponseExceptionAsync(final RetrySta
-                operationContext, shouldAcknowledge(batch, operationContext.getSessionContext()),
+                operationContext, shouldExpectResponse(batch, effectiveWriteConcern),
                 batch.getPayload(), batch.getFieldNameValidator());
     }
 
-    private void executeCommandAsync(final OperationContext operationContext, final AsyncConnection connection, final BulkWriteBatch batch,
+    private void executeCommandAsync(
+            final WriteConcern effectiveWriteConcern,
+            final OperationContext operationContext,
+            final AsyncConnection connection,
+            final BulkWriteBatch batch,
             final SingleResultCallback<BsonDocument> callback) {
         connection.commandAsync(namespace.getDatabaseName(), batch.getCommand(), NoOpFieldNameValidator.INSTANCE, null, batch.getDecoder(),
-                operationContext, shouldAcknowledge(batch, operationContext.getSessionContext()),
+                operationContext, shouldExpectResponse(batch, effectiveWriteConcern),
                 batch.getPayload(), batch.getFieldNameValidator(), callback);
     }
 
-    private WriteConcern getAppliedWriteConcern(final SessionContext sessionContext) {
-        if (sessionContext.hasActiveTransaction()) {
-            return WriteConcern.ACKNOWLEDGED;
-        } else {
-            return writeConcern;
+    private static WriteConcern validateAndGetEffectiveWriteConcern(final WriteConcern writeConcernSetting, final SessionContext sessionContext)
+            throws MongoClientException {
+        boolean activeTransaction = sessionContext.hasActiveTransaction();
+        WriteConcern effectiveWriteConcern = activeTransaction
+                ? WriteConcern.ACKNOWLEDGED
+                : writeConcernSetting;
+        if (sessionContext.hasSession() && !sessionContext.isImplicitSession() && !activeTransaction && !effectiveWriteConcern.isAcknowledged()) {
+            throw new MongoClientException("Unacknowledged writes are not supported when using an explicit session");
         }
+        return effectiveWriteConcern;
     }
 
-    private boolean shouldAcknowledge(final BulkWriteBatch batch, final SessionContext sessionContext) {
-        return ordered
-                ? batch.hasAnotherBatch() || getAppliedWriteConcern(sessionContext).isAcknowledged()
-                : getAppliedWriteConcern(sessionContext).isAcknowledged();
+    static Optional<WriteConcern> commandWriteConcern(final WriteConcern effectiveWriteConcern, final SessionContext sessionContext) {
+        return effectiveWriteConcern.isServerDefault() || sessionContext.hasActiveTransaction()
+                ? Optional.empty()
+                : Optional.of(effectiveWriteConcern);
+    }
+
+    private boolean shouldExpectResponse(final BulkWriteBatch batch, final WriteConcern effectiveWriteConcern) {
+        return effectiveWriteConcern.isAcknowledged() || (ordered && batch.hasAnotherBatch());
     }
```

Comment:
`shouldExpectResponse` replaces the `shouldAcknowledge` method. The new method has simpler body and named in a way that avoids erroneous allusions to an unacknowledged (`w: 0`) write concern.